docs/reference/programming_model.rst

```diff
@@ -0,0 +1,259 @@
+
+The hierarchy of threads introduced by :ref:`inherent_thread_model` is induced
+by the memory subsystem of GPUs. :numref:`memory_hierarchy` summarizes that memory namespaces and
+how they relate to the various levels of the threading model.
+
+.. _memory_hierarchy:
+
+.. figure:: ../data/reference/programming_model/memory_hierarchy.svg
+  :alt: Diagram depicting nested rectangles of varying color. The outermost one
+        titled "Grid", inside on the upper half a rectangle titled "Cluster".
+        Inside it are two identical rectangles titled "Block", inside them are
+        ones titled "Local" with multiple "Warp" titled rectangles. Blocks have
+        not just Local inside, but also rectangles titled "Shared". The Shared
+        rectangles of Blocks in the same Cluster are grouped together with a
+        translucent halo titled "Cluster shared". Outside the Cluster but
+        inside the Grid is a rectangle titled "Global" with three others
+        inside: "Surface", "Texture" (same color) and "Constant" (different
+        color).
+
+  Memory hierarchy.
+
+* Local or per-thread memory is read-write storage only visible to the
+  threads defining the given variables. The size of a block for a given kernel,
+  the number of concurrent warps are limited by local memory usage.
+  This relates to an important aspect: occupancy. This is the default memory
+  namespace.
+
+* Shared memory is read-write storage visible to all the threads in a given
+  block.
+
+* Distributed shared memory is read-write storage visible to all the threads
+  in a given block cluster.
+
+* Global memory is read-write storage visible to all threads in a given grid.
+  There are specialized versions of global memory with different usage
+  semantics which are typically backed by the same hardware storing global.
+
+  * Constant memory is read-only storage visible to all threads in a given
+    grid. It is a limited segment of global with queryable size.
+
+  * Texture memory is read-only storage visible to all threads in a given grid
+    and accessible through additional APIs.
+
+  * Surface is a writable version of texture memory.
+
+Execution Model
+===============================================================================
+
+HIP programs consist of two distinct scopes:
+
+* The host-side API running on the host processor. There are to APIs available:
+
+  * The HIP runtime API which enables use of the single-source programming
+    model.
+
+  * The HIP driver API which sits at a lower level and most importantly differs
+    by removing some of the facilities provided by the runtime API, most
+    importantly around kernel launching and argument setting. It is geared
+    towards implementing abstractions atop, such as the runtime API itself.
+
+* The device-side kernels running on GPUs. Both the host and the device-side
+  APIs have synchronous and asynchronous functions in them.
```

Comment:
Since this is a nested list of each "two" points (two distinct "scopes", two "APIs") I'd think making this a bit more visually appealing could help avoiding mixing this up. A definition list or a numbered list for the "outer" list could help.

And also changing "The host-side API running on the host processor" to just "The host-side running on the host processor", so that it won't be confused with the immediately following "two APIs available"

Also it doesn't really help, that the earlier mentioned "scopes" are actually "APIs", which might make it easier to confuse the nested lists.

Reply:
Slightly changed/reworded